### PR TITLE
Use ghc --merge-objs when available to join object files

### PR DESCRIPTION
**Pull request by bgamari**
Previously Cabal would rely on its own object-merge logic to produce
GHCi libraries. This was not ideal as it meant that this logic was
replicated in both GHC and Cabal. Moreover, Cabal's implementation
historically hasn't handled the more subtle aspects of interacting with
the linker, resulting in issues like #7828.

In GHC 9.4 we have introduced a new compiler mode, `--merge-objs`, which
exposes GHC's object merging mechanism. Here we teach Cabal to take
advantage of this mechanism when it is available.

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by Mikolaj** [comment]
Amazing! I suppose it doesn't need documenting? How about a tiny changelog file and a tiny test?

**Comment by bgamari** [comment]
It's technically an internal, non-user-visible change but I suppose a changelog entry won't hurt. I'm a bit unclear on what a test would look like; I would have thought that this functionality is exercised by nearly all of the "package tests" currently?

**Comment by Mikolaj** [comment]
Great, so I guess a minimal changelog file short suffice.

**Comment by Mikolaj** [comment]
I will write the changelog file, but please somebody review --- this is a tricky change.

**Comment by jneira** [comment]
the main concern could be we are not testing it for ghc > 9.2 but if it works for older ghcs in ci so the falback works i guess we will hit the possible bug in the other code branch sooner or later

or has been this tested in ghc itself?

**Comment by jneira** [comment]
> I will write the changelog file, but please somebody review --- this is a tricky change.

please do, it would help to move this forward

**Comment by Mikolaj** [comment]
I probably can't push to the branch, so here's the changelog (in the worst case I will merge it in a subsequent PR):
```
synopsis: Use 'ghc --merge-objs' whenever available, instead of internal cabal hacks, to join object files
packages: Cabal
issues: #7828
prs: #7960
```

**Comment by jneira** [comment]
@Mikolaj by default when you open a pr in github it allows upstream maintainers to do commits in the pr branch, you have to protect it on purpose so most pr branches are mofifiable but maintainers

**Review by jneira — APPROVED** [review]
lgtm, thanks @bgamari hope you dont mind i added the suggested changelog

**Comment by jneira** [comment]
@Mergifyio rebase master

**Comment by jneira** [comment]
It seems it closes #7828, at least for ghc >= 9.4

**Comment by bgamari** [comment]
Thank you, @jneira!

I have been holding off on pushing this MR forward since I have recently been engaged in a related set of work, moving GHC to use an LLVM-based toolchain on Windows, fixing a number of long-standing issues. Sadly, one of the compromises that we have had to make is to drop support for GHCi objects on Windows entirely due to deficiencies in the new (and old) linker. I have opened https://github.com/haskell/cabal/pull/8062 disabling support in Cabal.

As a result this MR is no longer needed as the problem that it solves will no longer exist with GHC 9.4 and later

**Comment by jneira** [comment]
@bgamari don't worry, it seemed to me than leverage that option would be good on its own, as ghc will know more about how to do the linking, as described in the initial comment
could not be this merge anyways to make more robust the ghc part?

**Comment by Mikolaj** [comment]
@bgamari: any harm if we merge this?

**Comment by Mikolaj** [comment]
@jneira: we decided, while this PR makes some sense, it's not strictly needed any more, and the stuff works as is, so it's safer not to touch it. However, if needed, e.g., if somebody has to change this bit of code, this PR is there.

**Pull request closed without merging**
